core/os/spin_lock.h

```diff
@@ -35,17 +35,46 @@
 
 #include <atomic>
 
+#if !(defined(__arm__) || defined(__aarch64__) || defined(_M_ARM) || defined(_M_ARM64))
+#if !defined(_MSC_VER)
+#include <immintrin.h>
+#endif
+#endif
+
+static_assert(std::atomic_bool::is_always_lock_free);
+
 class SpinLock {
-	mutable std::atomic_flag locked = ATOMIC_FLAG_INIT;
+	mutable std::atomic_bool locked;
 
 public:
 	_ALWAYS_INLINE_ void lock() const {
-		while (locked.test_and_set(std::memory_order_acquire)) {
-			// Continue.
+		while (true) {
+			bool expected = false;
+			if (!locked.compare_exchange_weak(expected, true, std::memory_order_acq_rel, std::memory_order_relaxed)) {
+				break;
+			}
+			while (locked.load(std::memory_order_relaxed)) {
+#if !defined(WEB_ENABLED)
+#if defined(__arm__) || defined(__aarch64__) || defined(_M_ARM) || defined(_M_ARM64)
+#ifdef _MSC_VER
+				__yield();
+#else
+				asm volatile("yield");
+#endif
+#else
+				_mm_pause();
```

Comment:
Rust seems to use `asm volatile(".insn i 0x0F, 0, x0, x0, 0x010")` on RISC-V (both 32 and 64) for the same purpose.

Reply:
Same in the Linux kernel https://github.com/torvalds/linux/blob/c2d5304e6c648ebcf653bace7e51e0e6742e46c8/arch/riscv/include/asm/vdso/processor.h#L25

And `asm volatile("or 27,27,27")` for PPC (https://stackoverflow.com/a/7588941), I do not think we have any support for other architectures at all (RISC-V is functional, PPC is at least in the supported arch list in the build scripts).